README.md
# SubRecon
Fast Subdomain Takeover Enumeration tool written in Go. It uses Golang concurrency and hence is very fast. It can easily detect potential subdomain takeovers that exist. Currently it only has 3 services, more services will be added soon..

## Installing

You need to have Golang installed on your machine. There are no additional requirements for this tool.

```sh
go get github.com/Asjidkalam/SubRecon
```

## Usage

` ./SubRecon -i hosts.txt -o results.txt`
- `-i` List of Subdomains
- `-o` Output/Result file


CHANGELOG.txt
SubRecon v0.0
=====================
This is the base/core version of the program.
the functions are working properly as intended.

DONE:
Base program working as intended.
takes hosts files as cli argument.
Service checks done with multiple goroutines.
Channels as waitgroups from goroutines.


TODO:
verbosity (hosts completed).
setting output file via cli argument.


SubRecon v0.1
==================
Changes:
Base program added with goroutine optimizations.

DONE:
(BUG FIX) Memleak fixed by implementing GarbageCollector after every loops on vulnHosts[]
(BUG FIX) Fixed output file writing.
Added lineCount function to count the total number of hosts from file
Increased Verbosity(total hosts completed) and formatting

TODO:
Enable concurrent HTTPS requests(Like a worker pool of 10-15 requests at a time).
	- Panic/Recover
Deploy checkService functions for each responses.
Banner and shit.


SubRecon v0.2
==================
Changes:
Fast as fuck!
Optimized the HTTP request concurrency.
Takes in a pool of requests concuurently.

TODO: 
Banner & formatting


test_env/v0.1/takeoverShit_v0.1.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strings"
)

/*
	Subdomain Takeover Automation
	Created by Asjid Kalam on 08/04/2020

	Currently supported takeover checks are from the following services:
	- Amazon S3 bucket
	- Readme.io
	- GitHub Pages
	...more services will be added soon.
*/
var vulnHosts [3]string

/*
	vulnHosts[ 0 => Amazon, 1 => GitHub Pages, 2 => Readme.io ]
*/

func checkAws(done chan bool, resp string, url string, status int) {
	if strings.Contains(resp, "NoSuchBucket") {
		fmt.Println("\n[>] Found Potential AWS S3 Takeover On", url)
		fmt.Println("[+] Status Code:", status)
		vulnHosts[0] = url
	}
	done <- true // channel for goroutine handling
}
func checkGpages(done chan bool, resp string, url string, status int) {
	if strings.Contains(resp, "There isn't a GitHub Pages site here.") {
		fmt.Println("\n[>] Found Potential GitHub Pages Takeover On", url)
		fmt.Println("[+] Status Code:", status)
		vulnHosts[1] = url
	}
	done <- true
}

func checkReadmeio(done chan bool, resp string, url string, status int) {
	if strings.Contains(resp, "Project doesnt exist... yet!") {
		fmt.Println("\n[>] Found Potential Readme.io Takeover On", url)
		fmt.Println("[+] Status Code:", status)
		vulnHosts[2] = url
	}
	done <- true
}

// this may not be the efficent way to line count
func lineCounter(r io.Reader, lineCounted chan int) {
	buf := make([]byte, 32*1024)
	count := 0
	lineSep := []byte{'\n'} // newline as separators

	for {
		c, err := r.Read(buf)
		count += bytes.Count(buf[:c], lineSep)

		switch {
		case err == io.EOF:
			lineCounted <- count

		case err != nil:
			fmt.Println("[!] Error occured: ", err)
		}
	}
}

func main() {
	var client http.Client

	fmt.Println("Subdomain Takeover Check")
	fmt.Println("==========================")

	if len(os.Args) != 3 {
		fmt.Println("Please specify the input and output file.")
		fmt.Println("Usage:", os.Args[0], "<hostsfile>.txt", "<outfile>.txt")
		return
	}
	hostFile := os.Args[1]
	outFile := os.Args[2]

	// declare unbuffered channels
	done := make(chan bool)
	lineCounted := make(chan int)

	file, err := os.Open(hostFile)
	if err != nil {
		log.Fatal("[!] Unable to read file", err)
	}
	defer file.Close()

	// experimental
	// create a io.Reader from the returned *os.File instance 'file'
	var reader io.Reader
	reader = file
	go lineCounter(reader, lineCounted)
	count := <-lineCounted
	running := 0
	fmt.Println("[+] Total Hosts:", count)

	// declaring file open again for Scanner
	file, err = os.Open(hostFile)
	if err != nil {
		log.Fatal("[!] Unable to read file", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// on each hosts..
	for scanner.Scan() {
		fmt.Printf("\n[*] (%d/%d) Hosts Completed", running, count)
		url := scanner.Text()
		if url != "" {
			// url format handling
			if !strings.Contains(url, "https://") && !strings.Contains(url, "http://") {
				url = "http://" + url
			}

			resp, err := client.Get(url)
			if err != nil {
				// do nothing
			}
			defer resp.Body.Close()

			// body parsing
			bodyBytes, err := ioutil.ReadAll(resp.Body)
			if err != nil {
				// do nothing
			}
			bodyString := string(bodyBytes)
			statusCode := int(resp.StatusCode)

			// setting output file
			outputFile, err := os.Create(outFile)
			if err != nil {
				fmt.Println("[!] Error while creating output file", err)
			}
			defer outputFile.Close()

			// deploy goroutine for checking each services concurrently
			go checkAws(done, bodyString, url, statusCode)
			go checkGpages(done, bodyString, url, statusCode)
			go checkReadmeio(done, bodyString, url, statusCode)

			// wait for goroutines to finish
			<-done
			<-done
			<-done

			// writing
			for i := 0; i < len(vulnHosts); i++ {
				if vulnHosts[i] != "" {
					if i == 0 {
						str := "[+] AWS: " + vulnHosts[i]
						outputFile.WriteString(str)
					}
					if i == 1 {
						str := "[+] GitHub Pages: " + vulnHosts[i]
						outputFile.WriteString(str)
					}
					if i == 2 {
						str := "[+] Readme.io: " + vulnHosts[i]
						outputFile.WriteString(str)
					}
					fmt.Println("[*] Output written to takeover_result.txt")
				}
			}
			// GC
			for i := 0; i < len(vulnHosts); i++ {
				vulnHosts[i] = ""
			}
		}
		running++
	}

	// we done.
	fmt.Println("\n[*] Takeover check completed.")

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}


SubRecon.go
package main

import (
	"bufio"
	"bytes"
	"crypto/tls"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"strings"
	"time"

	"./utils"
)

/*
	Subdomain Takeover Automation
	Created by Asjid Kalam on 08/04/2020

	Currently supported takeover checks are from the following services:
	- Amazon S3 bucket
	- Readme.io
	- GitHub Pages
	...more services will be added soon.
*/
var vulnHosts [3]string

// security checks
var tr = &http.Transport{
	TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	Dial: (&net.Dialer{
		Timeout:   15 * time.Second,
		KeepAlive: 15 * time.Second,
	}).Dial,
	TLSHandshakeTimeout: 10 * time.Second,
}

var client = &http.Client{
	Transport: tr,
	Timeout:   10 * time.Second,
}

//var wg sync.WaitGroup // handle sync

/*
	vulnHosts[ 0 => Amazon, 1 => GitHub Pages, 2 => Readme.io ]
*/
func checkAws(done chan bool, resp string, url string, status int) {
	if strings.Contains(resp, "NoSuchBucket") {
		fmt.Println("\n[>] Found Potential AWS S3 Takeover On", url)
		fmt.Println("[+] Status Code:", status)
		vulnHosts[0] = url
	}
	done <- true // channel for goroutine handling
}
func checkGpages(done chan bool, resp string, url string, status int) {
	if strings.Contains(resp, "There isn't a GitHub Pages site here.") {
		fmt.Println("\n[>] Found Potential GitHub Pages Takeover On", url)
		fmt.Println("[+] Status Code:", status)
		vulnHosts[1] = url
	}
	done <- true
}

func checkReadmeio(done chan bool, resp string, url string, status int) {
	if strings.Contains(resp, "Project doesnt exist... yet!") {
		fmt.Println("\n[>] Found Potential Readme.io Takeover On", url)
		fmt.Println("[+] Status Code:", status)
		vulnHosts[2] = url
	}
	done <- true
}

// this may not be the efficent way to line count
func lineCounter(r io.Reader, lineCounted chan int) {
	buf := make([]byte, 32*1024)
	count := 0
	lineSep := []byte{'\n'} // newline as separators

	for {
		c, err := r.Read(buf)
		count += bytes.Count(buf[:c], lineSep)

		switch {
		case err == io.EOF:
			lineCounted <- count

		case err != nil:
			fmt.Println("[!] Error occured: ", err)
		}
	}
}

// concurrent http requests
func MakeRequests(url string, bodyResp chan<- string, statusResp chan<- int) error {

	resp, err := client.Get(url)
	if err != nil {
		fmt.Println("[!]", err)
		return err
	}
	defer resp.Body.Close()

	// body parsing
	bodyBytes, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("[!]", err)
	}
	bodyResp <- string(bodyBytes)
	statusResp <- int(resp.StatusCode)

	//wg.Done()
	return nil
}

func main() {

	utils.Banner()

	// no verify
	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}

	hostFile := flag.String("i", "", ">> Hosts, input file")
	outFile := flag.String("o", "", ">> Output file")
	flag.Parse()

	if *hostFile == "" {
		fmt.Println(">> Usage: ./subrecon -i <hosts.txt> -o <results.txt>")
		os.Exit(1)
	}

	if *outFile == "" {
		fmt.Println(">> Please specify the output file.")
		os.Exit(1)
	}

	var url string

	// declare unbuffered channels
	done := make(chan bool)
	lineCounted := make(chan int)
	bodyResp := make(chan string)
	statusResp := make(chan int)

	file, err := os.Open(*hostFile)
	if err != nil {
		log.Fatal("[!] Unable to read file", err)
	}
	defer file.Close()

	// create a io.Reader from the returned *os.File instance 'file'
	var reader io.Reader
	reader = file
	go lineCounter(reader, lineCounted)
	count := <-lineCounted
	running := 0
	fmt.Println("[+] Total Hosts:", count)

	// declaring file open again for Scanner
	file, err = os.Open(*hostFile)
	if err != nil {
		log.Fatal("[!] Unable to read file", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	fmt.Println("[+] Please wait..")

	// on each hosts..
	for scanner.Scan() {
		url = scanner.Text()
		if url != "" {
			// url format handling
			if !strings.Contains(url, "https://") && !strings.Contains(url, "http://") {
				url = "http://" + url
			}
			// recurent calling of gouroutines here!
			//wg.Add(1)
			go MakeRequests(url, bodyResp, statusResp)
		}
	}

	//wg.Wait()
	// close(bodyResp)
	// close(statusResp)

	// output time.
	outputFile, err := os.Create(*outFile)
	if err != nil {
		fmt.Println("[!] Error while creating output file", err)
	}
	defer outputFile.Close()

	// check for return channels.
	for i := 0; i < count; i++ {
		bodyString := <-bodyResp
		statusCode := <-statusResp

		fmt.Printf("\n[*] (%d/%d) Hosts Completed", running, count)

		// deploy goroutine for checking each services concurrently
		go checkAws(done, bodyString, url, statusCode)
		go checkGpages(done, bodyString, url, statusCode)
		go checkReadmeio(done, bodyString, url, statusCode)

		// wait for goroutines to finish
		<-done
		<-done
		<-done

		// writing
		for i := 0; i < len(vulnHosts); i++ {
			if vulnHosts[i] != "" {
				fmt.Println("Writing")
				if i == 0 {
					str := "[+] AWS: " + vulnHosts[i]
					outputFile.WriteString(str)
				}
				if i == 1 {
					str := "[+] GitHub Pages: " + vulnHosts[i]
					outputFile.WriteString(str)
				}
				if i == 2 {
					str := "[+] Readme.io: " + vulnHosts[i]
					outputFile.WriteString(str)
				}
				fmt.Println("[*] Output written to", *outFile)
			}
		}
		// GC
		for i := 0; i < len(vulnHosts); i++ {
			vulnHosts[i] = ""
		}
		running++
	}

	// we done.
	fmt.Println("\n[*] Takeover check completed.")

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}
